daemon/hooks/hook_command.go

```diff
@@ -0,0 +1,142 @@
+package hooks
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/armon/circbuf"
+	"github.com/zrepl/zrepl/config"
+	"github.com/zrepl/zrepl/logger"
+)
+
+type HookEnv map[string]string
+
+func NewHookEnv(edge HookEdge, phase string, dryRun bool, extra map[string]string) HookEnv {
+	r := HookEnv{
+		"ZREPL_HOOKTYPE": edge.String(),
+		"ZREPL_PHASE":    phase,
+		"ZREPL_DRYRUN":   strconv.FormatBool(dryRun),
+	}
+
+	for k, v := range extra {
+		envName := fmt.Sprintf("ZREPL_%s", strings.ToUpper(k))
+		r[envName] = v
+	}
+
+	return r
+}
+
+type CommandHook struct {
+	edge       HookEdge
+	filter     config.FilesystemsFilter
+	errIsFatal bool
+	command    string
+	timeout    time.Duration
+}
+
+func NewCommandHook(in *config.HookCommand) (*CommandHook, error) {
+	r := &CommandHook{
+		filter:     in.Filesystems,
+		errIsFatal: in.ErrIsFatal,
+		command:    in.Command,
+		timeout:    in.Timeout,
+	}
+
+	switch in.Type {
+	case "command":
+		r.edge = All
+	case "pre-command":
+		r.edge = Pre
+	case "post-command":
+		r.edge = Post
+	default:
+		return nil, fmt.Errorf("unknown hook type: %s", in.Type)
+	}
+
+	return r, nil
+}
+
+func (h *CommandHook) Edge() HookEdge {
+	return h.edge
+}
+
+func (h *CommandHook) Filesystems() config.FilesystemsFilter {
+	return h.filter
+}
+
+func (h *CommandHook) ErrIsFatal() bool {
+	return h.errIsFatal
+}
+
+func (h *CommandHook) Run(ctx context.Context, edge HookEdge, phase string, dryRun bool, extra map[string]string) (runErr RunError) {
+	l := getLogger(ctx).WithField("command", h.command)
+
+	if !h.Edge().RunsDuring(edge) {
+		l.Debug(fmt.Sprintf("Run() called for wrong edge (edge=%s hook.Edge()=%s)",edge,h.Edge()))
+		return
+	}
+
+	cmdCtx, cancel := context.WithTimeout(ctx, h.timeout)
+	defer cancel()
+
+	// TODO: check for empty command string?
+	cmdExec := exec.CommandContext(cmdCtx, h.command)
+
+	hookEnv := NewHookEnv(edge, phase, dryRun, extra)
+	cmdEnv := os.Environ()
+	for k, v := range hookEnv {
+		cmdEnv = append(cmdEnv, fmt.Sprintf("%s=%s", k, v))
+	}
+	cmdExec.Env = cmdEnv
+
+	var scanMutex sync.Mutex
+	combinedOutput, err := circbuf.NewBuffer(int64(MAX_HOOK_LOG_SIZE))
+	if err != nil {
+		runErr.Err = err
+		return
+	}
+	logErrWriter := NewLogWriter(&scanMutex, l, logger.Warn, "stderr")
+	logOutWriter := NewLogWriter(&scanMutex, l, logger.Info, "stdout")
+	defer logErrWriter.Close()
+	defer logOutWriter.Close()
+
+	cmdExec.Stderr = io.MultiWriter(logErrWriter, combinedOutput)
+	cmdExec.Stdout = io.MultiWriter(logOutWriter, combinedOutput)
```

Comment:
combinedOutput (*circbuf.Buffer) is written to by both the stderr and stdout copier goroutine, but circbuf is not goroutine safe => race condition.

Reply:
I wasn't happy with circbuf.Buffer, as it required allocating too much memory up front, so I wrote a thread-safe `CircularLog` class that resizes the internal buffer as needed.

Reply:
If you are implementing a custom log buffer anyways, maybe try to copy the overflow handling from https://godoc.org/os/exec#ExitError 

>  // If the error output is long, Stderr may contain only a prefix
    // and suffix of the output, with the middle replaced with
    // text about the number of omitted bytes.

Not a hard requirement, however.
If possible, try to make it a sub-package of `./utils`, could imagine using it in package `./zfs` at some point.